### ISSUE TITLE
[BUG] Coredump after publishing paket

### ISSUE THREAD

**Issue by tbnobody** [bertmelis/espMqttClient#43]
**Describe the bug**
Today I got the following core dump:
```
09:49:54.696 > Guru Meditation Error: Core  1 panic'ed (StoreProhibited). Exception was unhandled.
09:49:54.696 > 
09:49:54.696 > Core  1 register dump:
09:49:54.696 > PC      : 0x400f4578  PS      : 0x00060c30  A0      : 0x800f45d0  A1      : 0x3ffb2650  
09:49:54.696 > A2      : 0x3ffd0ffc  A3      : 0x00000001  A4      : 0x3ffd15fc  A5      : 0x3ffd0d9c  
09:49:54.696 > A6      : 0x00000000  A7      : 0x00000004  A8      : 0x800f44f2  A9      : 0x3ffb2620  
09:49:54.696 > A10     : 0x00000000  A11     : 0x3ffb2687  A12     : 0x00000001  A13     : 0x3ffda2d0  
09:49:54.696 > A14     : 0x00000000  A15     : 0x3ffd0d9c  SAR     : 0x00000015  EXCCAUSE: 0x0000001d  
09:49:54.957 > EXCVADDR: 0x0000002c  LBEG    : 0x400898c8  LEND    : 0x400898d2  LCOUNT  : 0x00000000  
09:49:54.957 > 
09:49:54.957 > 
09:49:54.957 > Backtrace:0x400f4575:0x3ffb26500x400f45cd:0x3ffb26b0 0x400d813a:0x3ffb26d0 0x400d774d:0x3ffb2710 0x400e8734:0x3ffb2800 0x400fd735:0x3ffb2820 
09:49:54.957 > 
09:49:54.957 >   #0  0x400f4575:0x3ffb2650 in espMqttClientInternals::Outbox<espMqttClientInternals::Packet>::Iterator espMqttClientInternals::Outbox<espMqttClientInternals::Packet>::emplace<espMqttClientTypes::Error&, unsigned short&, char const*&, unsigned char const*&, unsigned int&, unsigned char&, bool&>(espMqttClientTypes::Error&, unsigned short&, char const*&, unsigned char const*&, unsigned int&, unsigned char&, bool&) at .pio/libdeps/generic/espMqttClient/src/Outbox.h:89
09:49:54.957 >       (inlined by) bool MqttClient::_addPacket<unsigned short&, char const*&, unsigned char const*&, unsigned int&, unsigned char&, bool&>(unsigned short&, char const*&, unsigned char const*&, unsigned int&, unsigned char&, bool&) at .pio/libdeps/generic/espMqttClient/src/MqttClient.h:149
09:49:54.957 >       (inlined by) MqttClient::publish(char const*, unsigned char, bool, unsigned char const*, unsigned int) at .pio/libdeps/generic/espMqttClient/src/MqttClient.cpp:152
09:49:54.957 >   #1  0x400f45cd:0x3ffb26b0 in MqttClient::publish(char const*, unsigned char, bool, char const*) at .pio/libdeps/generic/espMqttClient/src/MqttClient.cpp:163
09:49:54.957 >   #2  0x400d813a:0x3ffb26d0 in MqttSettingsClass::publish(String, String) at src/MqttSettings.cpp:240
09:49:54.957 >   #3  0x400d774d:0x3ffb2710 in MqttPublishingClass::loop() at src/MqttPublishing.cpp:43 (discriminator 9)
09:49:54.957 >   #4  0x400e8734:0x3ffb2800 in loop() at src/main.cpp:120
09:49:54.957 >   #5  0x400fd735:0x3ffb2820 in loopTask(void*) at C:/Users/Thomas/.platformio/packages/framework-arduinoespressif32/cores/esp32/main.cpp:50
```

Its a ESP32 without TLS and Platformio.
The following library versions are in use:

```
PACKAGES:
 - framework-arduinoespressif32 @ 3.20005.220925 (2.0.5)
 - tool-esptoolpy @ 1.40201.0 (4.2.1)
 - tool-openocd-esp32 @ 2.1100.20220706 (11.0)
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch3
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 43 compatible libraries
Scanning dependencies...
Dependency Graph
|-- ESP Async WebServer @ 1.2.3+sha.1dde9cf
|   |-- AsyncTCP @ 1.1.1
|   |-- ArduinoJson @ 6.19.4
|   |-- FS @ 2.0.0
|   |-- WiFi @ 2.0.0
|-- ArduinoJson @ 6.19.4
|-- espMqttClient @ 1.3.2+sha.f4c2f93
|   |-- AsyncTCP @ 1.1.1
|   |-- WiFiClientSecure @ 2.0.0
|   |   |-- WiFi @ 2.0.0
|   |-- WiFi @ 2.0.0
|-- RF24 @ 1.4.5
|   |-- SPI @ 2.0.0
|-- LittleFS @ 2.0.0
|   |-- FS @ 2.0.0
|-- Hoymiles @ 0.0.1
|   |-- RF24 @ 1.4.5
|   |   |-- SPI @ 2.0.0
|   |-- Every
|   |-- SPI @ 2.0.0
|   |-- CircularBuffer
|   |-- TimeoutHelper
|-- Ticker @ 2.0.0
|-- Ethernet @ 2.0.0
|   |-- WiFi @ 2.0.0
|-- WiFi @ 2.0.0
|-- Update @ 2.0.0
|-- ResetReason
|-- DNSServer @ 2.0.0
|   |-- WiFi @ 2.0.0
```

Currently I cannot reproduce the issue and it doesn't seem to occour often. In fact, it's the first time a saw it. But maybe you have an idee what could be wrong at Outbox.h:89 ?

**Comment by bertmelis** [comment]
I'm not at home ATM, will check this next week.

**Comment by bertmelis** [comment]
Do you have an idea about the state of the outbox when the crash happens? Empty, lots of packets, only one packet?

**Comment by bertmelis** [comment]
https://github.com/bertmelis/espMqttClient/blob/main/src/Outbox.h#L89

Deep inside the code, two possibilities:
- the queue is not empty but the pointer to the last element in the queue is not set.
- the queue is empty but the first element of the queue doesn't point to nullptr.

**Comment by bertmelis** [comment]
I think I found the culprit. when sending a PING request, the outbox wasn't guarded by the mutex. this probably caused the behaviour you experienced.

Data race issues are difficult to reproduce and in this case seemingly random.

Fixed with https://github.com/bertmelis/espMqttClient/commit/2caccea845f2333da1250354f0d58009b5a36566
Will issue a new release version soon.

**Comment by bertmelis** [comment]
Released with v1.3.3
